auditbeat/helper/hasher/hasher.go

```diff
@@ -0,0 +1,285 @@
+	"crypto/sha512"
+	"encoding/hex"
+	"fmt"
+	"hash"
+	"io"
+	"os"
+	"time"
+
+	"github.com/OneOfOne/xxhash"
+	"github.com/dustin/go-humanize"
+	"github.com/joeshaw/multierror"
+	"github.com/pkg/errors"
+	"golang.org/x/crypto/blake2b"
+	"golang.org/x/crypto/sha3"
+	"golang.org/x/time/rate"
+
+	"github.com/elastic/beats/libbeat/common/file"
+)
+
+// HashType identifies a cryptographic algorithm.
+type HashType string
+
+// Unpack unpacks a string to a HashType for config parsing.
+func (t *HashType) Unpack(v string) error {
+	*t = HashType(v)
+	return nil
+}
+
+// IsValid checks if the hash type is valid.
+func (t *HashType) IsValid() bool {
+	_, valid := validHashes[*t]
+	return valid
+}
+
+var validHashes = map[HashType]struct{}{
+	BLAKE2B_256: {},
+	BLAKE2B_384: {},
+	BLAKE2B_512: {},
+	MD5:         {},
+	SHA1:        {},
+	SHA224:      {},
+	SHA256:      {},
+	SHA384:      {},
+	SHA512:      {},
+	SHA512_224:  {},
+	SHA512_256:  {},
+	SHA3_224:    {},
+	SHA3_256:    {},
+	SHA3_384:    {},
+	SHA3_512:    {},
+	XXH64:       {},
+}
+
+// Enum of hash types.
+const (
+	BLAKE2B_256 HashType = "blake2b_256"
+	BLAKE2B_384 HashType = "blake2b_384"
+	BLAKE2B_512 HashType = "blake2b_512"
+	MD5         HashType = "md5"
+	SHA1        HashType = "sha1"
+	SHA224      HashType = "sha224"
+	SHA256      HashType = "sha256"
+	SHA384      HashType = "sha384"
+	SHA3_224    HashType = "sha3_224"
+	SHA3_256    HashType = "sha3_256"
+	SHA3_384    HashType = "sha3_384"
+	SHA3_512    HashType = "sha3_512"
+	SHA512      HashType = "sha512"
+	SHA512_224  HashType = "sha512_224"
+	SHA512_256  HashType = "sha512_256"
+	XXH64       HashType = "xxh64"
+)
+
+// Digest is a output of a hash function.
+type Digest []byte
+
+// String returns the digest value in lower-case hexadecimal form.
+func (d Digest) String() string {
+	return hex.EncodeToString(d)
+}
+
+// MarshalText encodes the digest to a hexadecimal representation of itself.
+func (d Digest) MarshalText() ([]byte, error) { return []byte(d.String()), nil }
+
+// FileTooLargeError is the error that occurs when a file that
+// exceeds the max file size is attempting to be hashed.
+type FileTooLargeError struct {
+	fileSize int64
+}
+
+// Error returns the error message for FileTooLargeError.
+func (e FileTooLargeError) Error() string {
+	return fmt.Sprintf("hasher: file size %d exceeds max file size", e.fileSize)
+}
+
+// Config contains the configuration of a FileHasher.
+type Config struct {
+	HashTypes           []HashType `config:"hash_types,replace"`
+	MaxFileSize         string     `config:"max_file_size"`
+	MaxFileSizeBytes    uint64     `config:",ignore"`
+	ScanRatePerSec      string     `config:"scan_rate_per_sec"`
+	ScanRateBytesPerSec uint64     `config:",ignore"`
+}
+
+// Validate validates the config.
+func (c *Config) Validate() error {
+	var errs multierror.Errors
+
+	for _, ht := range c.HashTypes {
+		if !ht.IsValid() {
+			errs = append(errs, errors.Errorf("invalid hash_types value '%v'", ht))
```

Comment:
Can we make it more error tolerant by lowercasing user input so Sha256 or SHA256 don't cause an error?

Reply:
Good idea, done. Very easy to do thanks to go-ucfg.